src/solve/canonicalization.md

```diff
@@ -1,10 +1,74 @@
+and reuse the result.
 
-[chalk]: https://rust-lang.github.io/chalk/book/canonical_queries/canonicalization.html#canonicalization
\ No newline at end of file
+Let's first go over the way canonical queries work and then dive into the specifics of
+how canonicalization works.
+
+## A walkthrough of canonical queries
+
+We're going to use the goal `u32: Trait<?x>` as an example and assume that this only holds
+for `u32: Trait<Vec<?z>>` where `?z` is unconstrained.
+
+### Canonicalizing the input
+
+We start by *canonicalizing* the goal, replacing inference variables with existential and
+placeholders with universal bound variables. This would result in the *canonical goal*
+`exists<T> u32: Trait<T>`.
+
+We remember the original values of all bound variables in the original context. Here this would
+map `T` back to `?x`. These original values are used later on when dealing with the query
+response.
+
+We now call the canonical query with the canonical goal.
+
+### Instantiating the canonical goal inside of the query
+
+To actually try to prove the canonical goal we start by instantiating existential variables with
+inference variables and universal variables without inference variables and placeholders again.
+
+This happens inside of the query in a completely separate `InferCtxt`. Inside of the query we
+now have a goal `u32: Trait<?0>`. We also remember which value we've used to instantiate the bound
+variables in the canonical goal, which maps `T` to `?0`.
+
+We now compute the goal `u32: Trait<?0>` and figure out that this holds, but we've constrained
+`?0` to `Vec<?1>`. We finally convert this result to something useful to the caller.
+
+### Canonicalizing the query response
+
+The caller has to know both the result of the query and any inference constraints from inside
+of the query. The result is already context independent, so we can return that as is, we do have
+to canonicalize the inference constraints though.
+
+For this we canonicalize the mapping from bound variables to the instantiated values in the query.
+This means that the query response is `Certainty::Yes` and a mapping from `T` to
+`exists<U> Vec<U>`.
+
+### Instantiating the query response
+
+The caller now has to apply the constraints returned by the query. For this they first
+instantiate the bound variables of the canonical response with inference variables and
+placeholders again, so the mapping in the response is now from `T` to `Vec<?z>`.
+
+It now equates the original value of `T` (`?x`) with the value for `T` in the
+response (`Vec<?z>`), which correctly constrains `?x` to `Vec<?z>`.
+
+## `ExternalConstraints`
+
+Computing a trait goal may not only constrain inference variables, it can also add region
+obligations, e.g. given a goal `(): AOutlivesB<'a, 'b>` we would like to return the fact that
+`'a: 'b` has to hold.
+
+This is done by not only returning the canonical `var_values` from the query but also extracting
```

Comment:
I dont think `var_values` has been mentioned yet, should be explicitly clear somewhere that the previously mentioned "mapping from `T` to `exists<U> Vec<U>`" is `var_values`

Reply:
Would it make sense to avoid bringing up the exact struct field names here and just talk about it in a high level way like the rest of this document? Maybe leaving the exact impl details to the "how exactly does canonicalization work"?